Objects/unicodeobject.c

```diff
@@ -2482,29 +2482,42 @@ unicode_fromformat_arg(_PyUnicodeWriter *writer,
         }
         assert(len >= 0);
 
-        if (precision < len)
-            precision = len;
+        int negative = buffer[0]=='-'?1:0;
+        len -= negative;
+
+        precision = Py_MAX(precision, len);
+        width = Py_MAX(width, precision + negative);
 
         arglen = Py_MAX(precision, width);
         if (_PyUnicodeWriter_Prepare(writer, arglen, 127) == -1)
             return NULL;
 
         if (width > precision) {
-            Py_UCS4 fillchar;
-            fill = width - precision;
-            fillchar = zeropad?'0':' ';
+            if (negative && zeropad) {
+                if (_PyUnicodeWriter_WriteChar(writer, '-') == -1)
+                    return NULL;
+            }
+
+            Py_UCS4 fillchar = zeropad?'0':' ';
+            fill = width - precision - negative;
             if (PyUnicode_Fill(writer->buffer, writer->pos, fill, fillchar) == -1)
                 return NULL;
             writer->pos += fill;
+
+            if (negative && !zeropad) {
+                if (_PyUnicodeWriter_WriteChar(writer, '-') == -1)
+                    return NULL;
+            }
```

Comment:
It can be simpler. Instead of writing minus explicitly, you can write it as a part of the buffer.

You will get the same result if remove this `if` and pass `&buffer[negative && zeropad]` instead of `&buffer[negative]` to `_PyUnicodeWriter_WriteASCIIString()`.

Reply:
That wouldn't work with precision: `%.5d` with `-123` would be `00-123`.